src/core/solvers/ipopt/multiple-shooting-nlp.cpp

```diff
@@ -0,0 +1,589 @@
+
+        for (size_t j = 0; j < ndx_; j++) {
+            g[i * ndx_ + j] = data_.x_diff[j];
+        }
+    }
+
+    // Initial conditions
+    data_.dx = Eigen::VectorXd::Map(x, ndx_);
+    state_->integrate(xs_[0], data_.dx, data_.x);
+
+    for (size_t j = 0; j < nx_; j++) {
+        g[T_ * ndx_ + j] = data_.x[j];
+    }
+
+    return true;
+}
+
+// [TNLP_eval_g]
+
+// [TNLP_eval_jac_g]
+// return the structure or values of the Jacobian
+bool MultipleShootingNlp::eval_jac_g(Ipopt::Index         n,
+                                     const Ipopt::Number *x,
+                                     bool                 new_x,
+                                     Ipopt::Index         m,
+                                     Ipopt::Index         nele_jac,
+                                     Ipopt::Index        *iRow,
+                                     Ipopt::Index        *jCol,
+                                     Ipopt::Number       *values)
+{
+    assert(n == nvar_);
+    assert(m == nconst_);
+
+    if (values == NULL) {
+        std::size_t idx_value = 0;
+        // Dynamic constraints
+        for (size_t idx_block = 0; idx_block < T_; idx_block++) {
+            for (size_t idx_row = 0; idx_row < ndx_; idx_row++) {
+                for (size_t idx_col = 0; idx_col < (2 * ndx_ + nu_); idx_col++) {
+                    iRow[idx_value] = idx_block * ndx_ + idx_row;
+                    jCol[idx_value] = idx_block * (ndx_ + nu_) + idx_col;
+                    idx_value++;
+                }
+            }
+        }
+
+        // Initial condition
+        for (size_t idx_row = 0; idx_row < nx_; idx_row++) {
+            for (size_t idx_col = 0; idx_col < ndx_; idx_col++) {
+                if (idx_row == idx_col) {
+                    iRow[idx_value] = T_ * ndx_ + idx_row;
+                    jCol[idx_value] = idx_col;
+                    idx_value++;
+                }
+            }
+        }
+
+        assert(nele_jac == idx_value);
+
+    } else {
+        std::size_t idx_value = 0;
+
+        // Dynamic constraints
+        for (size_t idx_block = 0; idx_block < T_; idx_block++) {
+            data_.dx      = Eigen::VectorXd::Map(x + idx_block * (ndx_ + nu_), ndx_);
+            data_.dxnext  = Eigen::VectorXd::Map(x + (idx_block + 1) * (ndx_ + nu_), ndx_);
+            data_.control = Eigen::VectorXd::Map(x + idx_block * (ndx_ + nu_) + ndx_, nu_);
+
+            state_->integrate(xs_[idx_block], data_.dx, data_.x);
+            state_->integrate(xs_[idx_block + 1], data_.dxnext, data_.xnext);
+
+            problem_->get_runningModels()[idx_block]->calc(problem_->get_runningDatas()[idx_block], data_.x,
+                                                           data_.control);
+            problem_->get_runningModels()[idx_block]->calcDiff(problem_->get_runningDatas()[idx_block], data_.x,
+                                                               data_.control);
```

Comment:
Same here.
We might not need to run `calc` and `calcDiff`.

Btw, you haven't run `calc` in the computation of cost derivatives!